File: data_baseline/openrouter__qwen__qwen3-coder-plus/scala/rep10/Main.scala
//> using scala "2.13"
//> using dep "org.http4s::http4s-dsl:0.23.34"
//> using dep "org.http4s::http4s-blaze-server:0.23.17"
//> using dep "org.http4s::http4s-circe:0.23.34"
//> using dep "io.circe::circe-generic:0.14.15"


import cats.effect._
import cats.implicits._
import org.http4s._
import org.http4s.dsl.Http4sDsl
import org.http4s.server.blaze.BlazeServerBuilder
import org.http4s.headers.`Set-Cookie`
import org.http4s.HttpRoutes
import org.http4s.implicits._
import org.http4s.circe._
import io.circe._
import io.circe.generic.semiauto._
import io.circe.syntax._

import java.time.Instant
import java.time.format.DateTimeFormatter  
import java.util.UUID
import scala.collection.mutable
import scala.util.matching.Regex

case class User(id: Int, username: String, passwordHash: String)

object User {
  implicit val userEncoder: Encoder[User] = deriveEncoder[User]
  implicit val userDecoder: Decoder[User] = deriveDecoder[User].emap { user =>
    // Don't expose password hash in response  
    Right(user.copy(passwordHash = ""))
  }
}

case class Todo(
  id: Int,
  title: String,
  description: String,
  completed: Boolean,
  created_at: String,
  updated_at: String
)

object Todo {
  implicit val todoEncoder: Encoder[Todo] = deriveEncoder[Todo]
  implicit val todoDecoder: Decoder[Todo] = deriveDecoder[Todo]
}

object DateTimeUtil {
  def now(): String = Instant.now().toString.replaceAll("\\[.*?\\]", "").take(19) + "Z"
}

class TodoApp extends Http4sDsl[IO] {

  // In-memory storage
  private var userCounter = 0
  private val users = mutable.Map[String, User]() // key: username
  private val sessionToUser = mutable.Map[String, Int]() // key: session_id -> value: user_id
  private var todoCounter = 0
  private val todos = mutable.Map[Int, mutable.Map[Int, Todo]]() // key: user_id -> todo_id -> todo

  // Helper to get next user ID
  private def getNextUserId(): Int = {
    userCounter += 1
    userCounter
  }

  // Helper to get next todo ID
  private def getNextTodoId(): Int = {
    todoCounter += 1
    todoCounter
  }

  // Helper to validate username format
  private val usernamePattern: Regex = "^[a-zA-Z0-9_]+$".r
  private def isValidUsername(username: String): Boolean = {
    username.length >= 3 && 
    username.length <= 50 && 
    usernamePattern.pattern.matcher(username).matches()
  }

  // Helper to validate password (minimum length)
  private def isValidPassword(password: String): Boolean = password.length >= 8

  // Helper to hash password (using simple approach for demo)
  private def hashPassword(password: String): String = password.reverse // In real app, use proper hashing like bcrypt

  // Helper to generate session ID
  private def generateSessionId(): String = UUID.randomUUID().toString.replace("-", "")

  // Authenticate based on session cookie
  private def authenticate(req: Request[IO]): Option[Int] = {
    req.cookies.find(_.name == "session_id") match {
      case Some(cookie) =>
        sessionToUser.get(cookie.content)
      case None => None
    }
  }

  // Middleware to check authentication
  private def withAuth(request: Request[IO])(block: Int => IO[Response[IO]]): IO[Response[IO]] = {
    authenticate(request) match {
      case Some(userId) => block(userId)
      case None => 
        Response[IO](Status.Unauthorized).withEntity(Json.obj("error" := "Authentication required")).pure[IO]
    }
  }

  // JSON decoders and encoders for the routes
  implicit val jsonDecoder: EntityDecoder[IO, Json] = jsonOf[IO, Json]

  // Register endpoint
  val registerRoute: HttpRoutes[IO] = HttpRoutes.of[IO] {
    case req @ POST -> Root / "register" =>
      req.attemptAs[Json].value.flatMap {
        case Right(json) =>
          (json.hcursor.downField("username").as[String], json.hcursor.downField("password").as[String]) match {
            case (Right(username), Right(password)) =>
              if (!isValidUsername(username)) {
                BadRequest(Json.obj("error" := "Invalid username"))
              } else if (!isValidPassword(password)) {
                BadRequest(Json.obj("error" := "Password too short"))
              } else if (users.contains(username)) {
                Conflict(Json.obj("error" := "Username already exists"))
              } else {
                val userId = getNextUserId()
                val user = User(userId, username, hashPassword(password))
                users(username) = user
                Created(user.asJson)
              }
            case _ => BadRequest(Json.obj("error" := "Invalid request body"))
          }
        case Left(_) => BadRequest(Json.obj("error" := "Invalid JSON"))
      }
  }

  // Login endpoint
  val loginRoute: HttpRoutes[IO] = HttpRoutes.of[IO] {
    case req @ POST -> Root / "login" =>
      req.attemptAs[Json].value.flatMap {
        case Right(json) =>
          (json.hcursor.downField("username").as[String], json.hcursor.downField("password").as[String]) match {
            case (Right(inputUsername), Right(inputPassword)) =>
              users.get(inputUsername) match {
                case Some(user) if user.passwordHash == hashPassword(inputPassword) =>
                  val sessionId = generateSessionId()
                  sessionToUser(sessionId) = user.id
                  val cookie = ResponseCookie(
                    name = "session_id",
                    content = sessionId,
                    path = Some("/"),
                    httpOnly = true
                  )
                  Ok(user.asJson).map(_.addCookie(cookie))
                case _ => Response[IO](Status.Unauthorized).withEntity(Json.obj("error" := "Invalid credentials")).pure[IO]
              }
            case _ => BadRequest(Json.obj("error" := "Invalid request body"))
          }
        case Left(_) => BadRequest(Json.obj("error" := "Invalid JSON"))
      }
  }

  // Logout endpoint
  val logoutRoute: HttpRoutes[IO] = HttpRoutes.of[IO] {
    case req @ POST -> Root / "logout" =>
      withAuth(req) { userId =>
        req.cookies.find(_.name == "session_id").foreach { cookie =>
          sessionToUser.remove(cookie.content)
        }
        Ok(Json.obj())
      }
  }

  // Get current user endpoint
  val getMeRoute: HttpRoutes[IO] = HttpRoutes.of[IO] {
    case req @ GET -> Root / "me" =>
      withAuth(req) { userId =>
        users.values.find(_.id == userId) match {
          case Some(user) => 
            // Don't include password hash in response
            val responseUser = User(user.id, user.username, "")  
            Ok(responseUser.asJson)
          case None => Response[IO](Status.Unauthorized).withEntity(Json.obj("error" := "Authentication required")).pure[IO]
        }
      }
  }

  // Update password endpoint
  val updatePasswordRoute: HttpRoutes[IO] = HttpRoutes.of[IO] {
    case req @ PUT -> Root / "password" =>
      withAuth(req) { userId =>
        req.attemptAs[Json].value.flatMap {
          case Right(json) =>
            (json.hcursor.downField("old_password").as[String], json.hcursor.downField("new_password").as[String]) match {
              case (Right(oldPassword), Right(newPassword)) =>
                users.values.find(_.id == userId) match {
                  case Some(user) if user.passwordHash == hashPassword(oldPassword) =>
                    if (!isValidPassword(newPassword)) {
                      BadRequest(Json.obj("error" := "Password too short"))
                    } else {
                      val updatedUser = user.copy(passwordHash = hashPassword(newPassword))
                      users(user.username) = updatedUser
                      Ok(Json.obj())
                    }
                  case _ => Response[IO](Status.Unauthorized).withEntity(Json.obj("error" := "Invalid credentials")).pure[IO]
                }
              case _ => BadRequest(Json.obj("error" := "Invalid request body"))  
            }
          case Left(_) => BadRequest(Json.obj("error" := "Invalid JSON"))
        }
      }
  }
  
  // Initialize user's todo map
  private def ensureUserTodosExist(userId: Int): Unit = {
    if (!todos.contains(userId)) {
      todos(userId) = mutable.Map[Int, Todo]()
    }
  }

  // List todos for user
  val getTodosRoute: HttpRoutes[IO] = HttpRoutes.of[IO] {
    case req @ GET -> Root / "todos" =>
      withAuth(req) { userId =>
        ensureUserTodosExist(userId)
        val userTodos = todos(userId).values.toList.sortBy(_.id)
        Ok(userTodos.asJson)
      }
  }

  // Create a new todo
  val createTodoRoute: HttpRoutes[IO] = HttpRoutes.of[IO] {
    case req @ POST -> Root / "todos" =>
      withAuth(req) { userId =>
        req.attemptAs[Json].value.flatMap {
          case Right(json) =>
            val titleOpt = json.hcursor.downField("title").as[String].toOption
            val descOpt = json.hcursor.downField("description").as[String].toOption.orElse(Some(""))
            
            titleOpt match {
              case Some(title) if title.nonEmpty =>
                val timestamp = DateTimeUtil.now()
                val newId = getNextTodoId()
                val todo = Todo(newId, title, descOpt.getOrElse(""), false, timestamp, timestamp)
                
                ensureUserTodosExist(userId)
                todos(userId)(newId) = todo
                Created(todo.asJson)
              case _ => BadRequest(Json.obj("error" := "Title is required"))
            }
          case Left(_) => BadRequest(Json.obj("error" := "Invalid JSON"))
        }
      }
  }

  // Get a specific todo
  val getTodoByIdRoute: HttpRoutes[IO] = HttpRoutes.of[IO] {
    case req @ GET -> Root / "todos" / IntVar(todoId) =>
      withAuth(req) { userId =>
        ensureUserTodosExist(userId)
        todos(userId).get(todoId) match {
          case Some(todo) => Ok(todo.asJson)
          case None => Response[IO](Status.NotFound).withEntity(Json.obj("error" := "Todo not found")).pure[IO]
        }
      }
  }

  // Update a specific todo (partial update)
  val updateTodoRoute: HttpRoutes[IO] = HttpRoutes.of[IO] {
    case req @ PUT -> Root / "todos" / IntVar(todoId) =>
      withAuth(req) { userId =>
        ensureUserTodosExist(userId)
        todos(userId).get(todoId) match {
          case Some(existingTodo) =>
            req.attemptAs[Json].value.flatMap {
              case Right(json) =>
                val updatedTitle = json.hcursor.downField("title").as[String].toOption.getOrElse(existingTodo.title)
                val updatedDesc = json.hcursor.downField("description").as[String].toOption.getOrElse(existingTodo.description)
                val updatedCompleted = json.hcursor.downField("completed").as[Boolean].toOption.getOrElse(existingTodo.completed)
                
                if (updatedTitle.isEmpty) {
                  BadRequest(Json.obj("error" := "Title is required"))
                } else {
                  val timestamp = DateTimeUtil.now()
                  val updatedTodo = existingTodo.copy(
                    title = updatedTitle,
                    description = updatedDesc,
                    completed = updatedCompleted,
                    updated_at = timestamp
                  )
                  
                  todos(userId)(todoId) = updatedTodo
                  Ok(updatedTodo.asJson)
                }
              case Left(_) => BadRequest(Json.obj("error" := "Invalid JSON"))
            }
          case None => Response[IO](Status.NotFound).withEntity(Json.obj("error" := "Todo not found")).pure[IO]
        }
      }
  }

  // Delete a specific todo
  val deleteTodoRoute: HttpRoutes[IO] = HttpRoutes.of[IO] {
    case req @ DELETE -> Root / "todos" / IntVar(todoId) =>
      withAuth(req) { userId =>
        ensureUserTodosExist(userId)
        if (todos(userId).contains(todoId)) {
          todos(userId).remove(todoId)
          Response[IO](Status.NoContent).pure[IO]
        } else {
          Response[IO](Status.NotFound).withEntity(Json.obj("error" := "Todo not found")).pure[IO]
        }
      }
  }

  val routes: HttpRoutes[IO] = registerRoute <+> 
                              loginRoute <+> 
                              logoutRoute <+> 
                              getMeRoute <+> 
                              updatePasswordRoute <+> 
                              getTodosRoute <+> 
                              createTodoRoute <+> 
                              getTodoByIdRoute <+> 
                              updateTodoRoute <+> 
                              deleteTodoRoute
}

object Main extends IOApp.Simple {
  def run: IO[Unit] = {
    import sys.process._
    
    val args = System.getProperty("scala.cli.args", "").split(" ")
    val portIdx = args.indexOf("--port")
    val port = if (portIdx >= 0) {
      try args(portIdx + 1).toInt catch { case _: Exception => 8080 }
    } else {
      8080
    }
    
    val httpApp = new TodoApp()
    
    BlazeServerBuilder[IO]
      .bindHttp(port, "0.0.0.0") // Bind to all interfaces
      .withHttpApp(httpApp.routes.orNotFound)
      .serve
      .compile
      .drain
  }
}

File: data_baseline/openrouter__qwen__qwen3-coder-plus/rust/rep10/src/main.rs
use actix_web::{web, App, HttpResponse, HttpServer, Result, middleware, cookie::Cookie};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use chrono::{Utc, DateTime};
use regex::Regex;

#[derive(Debug, Clone, Serialize, Deserialize)]
struct User {
    id: u32,
    username: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct Todo {
    id: u32,
    title: String,
    description: String,
    completed: bool,
    created_at: DateTime<Utc>,
    updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct NewTodo {
    title: String,
    description: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct UpdateTodo {
    title: Option<String>,
    description: Option<String>,
    completed: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct RegisterData {
    username: String,
    password: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct LoginData {
    username: String,
    password: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct PasswordUpdateData {
    old_password: String,
    new_password: String,
}

// Application state: stores users, todos, sessions and passwords
struct AppState {
    users: Arc<Mutex<HashMap<u32, User>>>,
    todos: Arc<Mutex<HashMap<u32, HashMap<u32, Todo>>>>, // user_id -> todo_id -> Todo
    sessions: Arc<Mutex<HashMap<String, u32>>>, // session_token -> user_id
    passwords: Arc<Mutex<HashMap<u32, String>>>, // user_id -> password
    next_user_id: Arc<Mutex<u32>>,
    next_todo_id: Arc<Mutex<u32>>,
}

fn json_response<T: Serialize>(data: T) -> Result<HttpResponse> {
    Ok(HttpResponse::Ok()
        .content_type("application/json")
        .json(data))
}

fn json_error(status: actix_web::http::StatusCode, message: &str) -> Result<HttpResponse> {
    Ok(HttpResponse::build(status)
        .content_type("application/json")
        .json(serde_json::json!({"error": message})))
}

fn auth_check(req: &actix_web::HttpRequest, sessions: &Arc<Mutex<HashMap<String, u32>>>) -> Option<u32> {
    if let Some(cookie) = req.cookie("session_id") {
        let session_token = cookie.value();
        let sessions_lock = sessions.lock().unwrap();
        if let Some(user_id) = sessions_lock.get(session_token).cloned() {
            return Some(user_id);
        }
    }
    None
}

fn validate_username(username: &str) -> bool {
    if username.len() < 3 || username.len() > 50 {
        return false;
    }
    Regex::new(r"^[a-zA-Z0-9_]+$").unwrap().is_match(username)
}

async fn register(
    data: web::Json<RegisterData>,
    app_state: web::Data<AppState>,
) -> Result<HttpResponse> {
    if !validate_username(&data.username) {
        return json_error(actix_web::http::StatusCode::BAD_REQUEST, "Invalid username");
    }

    if data.password.len() < 8 {
        return json_error(actix_web::http::StatusCode::BAD_REQUEST, "Password too short");
    }

    let mut users = app_state.users.lock().unwrap();
    
    // Check if username already exists
    for (_, user) in users.iter() {
        if user.username == data.username {
            return json_error(actix_web::http::StatusCode::CONFLICT, "Username already exists");
        }
    }

    let user_id = *app_state.next_user_id.lock().unwrap();
    *app_state.next_user_id.lock().unwrap() += 1;

    let user = User {
        id: user_id,
        username: data.username.clone(),
    };

    users.insert(user_id, user);

    let mut passwords = app_state.passwords.lock().unwrap();
    passwords.insert(user_id, data.password.clone());

    let user_todos_map = HashMap::new();
    app_state.todos.lock().unwrap().insert(user_id, user_todos_map);

    let response = User {
        id: user_id,
        username: data.username.clone(),
    };

    Ok(HttpResponse::Created()
        .content_type("application/json")
        .json(response))
}

async fn login(
    data: web::Json<LoginData>,
    app_state: web::Data<AppState>,
) -> Result<HttpResponse> {
    let users = app_state.users.lock().unwrap();
    let mut user_id_opt = None;
    
    for (id, user) in users.iter() {
        if user.username == data.username {
            let passwords = app_state.passwords.lock().unwrap();
            if let Some(password) = passwords.get(id) {
                if password == &data.password {
                    user_id_opt = Some(*id);
                    break;
                }
            }
        }
    }

    if let Some(user_id) = user_id_opt {
        let session_id = uuid::Uuid::new_v4().to_string();
        
        {
            let mut sessions = app_state.sessions.lock().unwrap();
            sessions.insert(session_id.clone(), user_id);
        }
        
        let _cookies = app_state.sessions.lock().unwrap();
        let response_data = User {
            id: user_id,
            username: data.username.clone(),
        };
        
        let mut cookie = Cookie::new("session_id", &session_id);
        cookie.set_path("/");
        cookie.set_http_only(true);
        
        Ok(HttpResponse::Ok()
            .content_type("application/json")
            .cookie(cookie)
            .json(response_data))
    } else {
        json_error(actix_web::http::StatusCode::UNAUTHORIZED, "Invalid credentials")
    }
}

async fn logout(
    req: actix_web::HttpRequest,
    app_state: web::Data<AppState>,
) -> Result<HttpResponse> {
    let sessions = app_state.sessions.clone();
    let user_id = auth_check(&req, &sessions);
    
    if user_id.is_none() {
        return json_error(actix_web::http::StatusCode::UNAUTHORIZED, "Authentication required");
    }

    // Remove the session token server-side
    if let Some(cookie) = req.cookie("session_id") {
        let session_token = cookie.value();
        let mut sessions_lock = app_state.sessions.lock().unwrap();
        sessions_lock.remove(session_token);
    }

    Ok(HttpResponse::Ok()
        .content_type("application/json")
        .json(serde_json::json!({})))
}

async fn get_user_info(
    req: actix_web::HttpRequest,
    app_state: web::Data<AppState>,
) -> Result<HttpResponse> {
    let sessions = app_state.sessions.clone();
    let user_id = auth_check(&req, &sessions);
    
    if user_id.is_none() {
        return json_error(actix_web::http::StatusCode::UNAUTHORIZED, "Authentication required");
    }

    let users = app_state.users.lock().unwrap();
    if let Some(user) = users.get(&user_id.unwrap()) {
        return json_response(User {
            id: user.id,
            username: user.username.clone(),
        });
    }

    json_error(actix_web::http::StatusCode::UNAUTHORIZED, "Authentication required")
}

async fn update_password(
    req: actix_web::HttpRequest,
    data: web::Json<PasswordUpdateData>,
    app_state: web::Data<AppState>,
) -> Result<HttpResponse> {
    let sessions = app_state.sessions.clone();
    let user_id = auth_check(&req, &sessions);
    
    if user_id.is_none() {
        return json_error(actix_web::http::StatusCode::UNAUTHORIZED, "Authentication required");
    }

    let user_id = user_id.unwrap();

    // Check if old password matches
    let passwords = app_state.passwords.lock().unwrap();
    let current_password = passwords.get(&user_id).unwrap();

    if current_password != &data.old_password {
        return json_error(actix_web::http::StatusCode::UNAUTHORIZED, "Invalid credentials");
    }

    if data.new_password.len() < 8 {
        return json_error(actix_web::http::StatusCode::BAD_REQUEST, "Password too short");
    }

    drop(passwords);
    
    let mut passwords = app_state.passwords.lock().unwrap();
    passwords.insert(user_id, data.new_password.clone());

    Ok(HttpResponse::Ok()
        .content_type("application/json")
        .json(serde_json::json!({})))
}

async fn get_todos(
    req: actix_web::HttpRequest,
    app_state: web::Data<AppState>,
) -> Result<HttpResponse> {
    let sessions = app_state.sessions.clone();
    let user_id = auth_check(&req, &sessions);
    
    if user_id.is_none() {
        return json_error(actix_web::http::StatusCode::UNAUTHORIZED, "Authentication required");
    }

    let user_id = user_id.unwrap();
    
    let todos = app_state.todos.lock().unwrap();
    let user_todos = todos.get(&user_id).unwrap();
    
    let mut todos_list: Vec<Todo> = user_todos.values().cloned().collect();
    todos_list.sort_by(|a, b| a.id.cmp(&b.id));  // Sort by id ascending

    json_response(todos_list)
}

async fn create_todo(
    req: actix_web::HttpRequest,
    data: web::Json<NewTodo>,
    app_state: web::Data<AppState>,
) -> Result<HttpResponse> {
    let sessions = app_state.sessions.clone();
    let user_id = auth_check(&req, &sessions);
    
    if user_id.is_none() {
        return json_error(actix_web::http::StatusCode::UNAUTHORIZED, "Authentication required");
    }

    if data.title.is_empty() {
        return json_error(actix_web::http::StatusCode::BAD_REQUEST, "Title is required");
    }

    let todo_id = *app_state.next_todo_id.lock().unwrap();
    *app_state.next_todo_id.lock().unwrap() += 1;

    let now = Utc::now();

    let todo = Todo {
        id: todo_id,
        title: data.title.clone(),
        description: data.description.as_ref().cloned().unwrap_or_else(|| "".to_string()),
        completed: false,
        created_at: now,
        updated_at: now,
    };

    let mut todos = app_state.todos.lock().unwrap();
    let user_todos = todos.get_mut(&user_id.unwrap()).unwrap();
    user_todos.insert(todo_id, todo);

    let response = todos.get(&user_id.unwrap()).unwrap().get(&todo_id).unwrap();

    Ok(HttpResponse::Created()
        .content_type("application/json")
        .json(response))
}

async fn get_todo(
    req: actix_web::HttpRequest,
    path: web::Path<u32>,
    app_state: web::Data<AppState>,
) -> Result<HttpResponse> {
    let sessions = app_state.sessions.clone();
    let user_id = auth_check(&req, &sessions);
    
    if user_id.is_none() {
        return json_error(actix_web::http::StatusCode::UNAUTHORIZED, "Authentication required");
    }

    let todo_id = path.into_inner();
    
    let todos = app_state.todos.lock().unwrap();
    let user_todos = todos.get(&user_id.unwrap());
    
    if let Some(user_todos) = user_todos {
        if let Some(todo) = user_todos.get(&todo_id) {
            return json_response(todo);
        }
    }
    
    json_error(actix_web::http::StatusCode::NOT_FOUND, "Todo not found")
}

async fn update_todo(
    req: actix_web::HttpRequest,
    path: web::Path<u32>,
    data: web::Json<UpdateTodo>,
    app_state: web::Data<AppState>,
) -> Result<HttpResponse> {
    let sessions = app_state.sessions.clone();
    let user_id = auth_check(&req, &sessions);
    
    if user_id.is_none() {
        return json_error(actix_web::http::StatusCode::UNAUTHORIZED, "Authentication required");
    }

    let user_id = user_id.unwrap();
    let todo_id = path.into_inner();

    if let Some(ref title) = data.title {
        if title.is_empty() {
            return json_error(actix_web::http::StatusCode::BAD_REQUEST, "Title is required");
        }
    }

    let mut todos = app_state.todos.lock().unwrap();
    
    if let Some(user_todos) = todos.get_mut(&user_id) {
        if let Some(todo) = user_todos.get_mut(&todo_id) {
            // Update fields if they're provided
            if let Some(title) = &data.title {
                todo.title = title.clone();
            }
            if let Some(description) = &data.description {
                todo.description = description.clone();
            }
            if let Some(completed) = data.completed {
                todo.completed = completed;
            }
            
            // Update timestamps
            todo.updated_at = Utc::now();
            
            let updated_todo = user_todos.get(&todo_id).unwrap().clone();
            return json_response(updated_todo);
        }
    }
    
    json_error(actix_web::http::StatusCode::NOT_FOUND, "Todo not found")
}

async fn delete_todo(
    req: actix_web::HttpRequest,
    path: web::Path<u32>,
    app_state: web::Data<AppState>,
) -> Result<HttpResponse> {
    let sessions = app_state.sessions.clone();
    let user_id = auth_check(&req, &sessions);
    
    if user_id.is_none() {
        return json_error(actix_web::http::StatusCode::UNAUTHORIZED, "Authentication required");
    }

    let user_id = user_id.unwrap();
    let todo_id = path.into_inner();

    let mut todos = app_state.todos.lock().unwrap();
    if let Some(user_todos) = todos.get_mut(&user_id) {
        if user_todos.contains_key(&todo_id) {
            user_todos.remove(&todo_id);
            return Ok(HttpResponse::NoContent().finish());
        }
    }
    
    json_error(actix_web::http::StatusCode::NOT_FOUND, "Todo not found")
}

#[actix_web::main]
async fn main() -> std::io::Result<()> {
    let args: Vec<String> = std::env::args().collect();
    
    let mut port = 8080; // default port
    for i in 0..args.len() {
        if args[i] == "--port" && i + 1 < args.len() {
            port = args[i + 1].parse::<u16>().expect("Invalid port number");
            break;
        }
    }

    let state = web::Data::new(AppState {
        users: Arc::new(Mutex::new(HashMap::new())),
        todos: Arc::new(Mutex::new(HashMap::new())),
        sessions: Arc::new(Mutex::new(HashMap::new())),
        passwords: Arc::new(Mutex::new(HashMap::new())),
        next_user_id: Arc::new(Mutex::new(1)),
        next_todo_id: Arc::new(Mutex::new(1)),
    });

    println!("Starting server on 0.0.0.0:{}", port);

    HttpServer::new(move || {
        App::new()
            .app_data(state.clone())
            .wrap(middleware::Logger::default())
            .route("/register", web::post().to(register))
            .route("/login", web::post().to(login))
            .route("/logout", web::post().to(logout))
            .route("/me", web::get().to(get_user_info))
            .route("/password", web::put().to(update_password))
            .route("/todos", web::get().to(get_todos))
            .route("/todos", web::post().to(create_todo))
            .route("/todos/{id}", web::get().to(get_todo))
            .route("/todos/{id}", web::put().to(update_todo))
            .route("/todos/{id}", web::delete().to(delete_todo))
    })
    .bind(("0.0.0.0", port))?
    .run()
    .await
}